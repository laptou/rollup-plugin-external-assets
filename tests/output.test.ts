import { outputSnapshotMacro } from "./macros";
import externalAssets from "../src/index";

describe("output formats", () => {
	const outputFormats = ["es", "cjs", "amd", "iife", "umd", "system"] as const;

	for (const format of outputFormats) {
		test(format, () => {
			return outputSnapshotMacro(
				{
					input: "tests/fixtures/src/index1.js",
					plugins: [
						externalAssets("tests/fixtures/assets/*"),
					],
				},
				{
					name: "mybundle", // required for iife bundles.
					format,
				}
			);
		});
	}
});

test("output.sourcemaps = true", () => {
	return outputSnapshotMacro(
		{
			input: "tests/fixtures/src/index1.js",
			plugins: [
				externalAssets("tests/fixtures/assets/*"),
			],
		},
		{
			sourcemap: true,
		}
	);
});

test("output.dir = 'sub/abcd'", () => {
	return outputSnapshotMacro(
		{
			input: "tests/fixtures/src/index1.js",
			plugins: [
				externalAssets("tests/fixtures/assets/*"),
			],
		},
		{
			dir: "sub/abcd",
		}
	);
});

test(
	[
		"output.dir = 'sub/abcd'",
		"output.entryFileNames = 'a/b/[name].js'",
	].join(" && "),
	() => {
		return outputSnapshotMacro(
			{
				input: "tests/fixtures/src/index1.js",
				plugins: [
					externalAssets("tests/fixtures/assets/*"),
				],
			},
			{
				dir: "sub/abcd",
				entryFileNames: "a/b/[name].js",
			}
		);
	}
);

test(
	[
		"output.dir = 'sub/abcd'",
		"output.entryFileNames = 'a/b/[name].js'",
		"output.assetFileNames = 'assets/sub/[name]_[hash][extname]'",
	].join(" && "),
	() => {
		return outputSnapshotMacro(
			{
				input: "tests/fixtures/src/index1.js",
				plugins: [
					externalAssets("tests/fixtures/assets/*"),
				],
			},
			{
				dir: "sub/abcd",
				entryFileNames: "a/b/[name].js",
				assetFileNames: "assets/sub/[name]_[hash][extname]",
			}
		);
	}
);

test(
	[
		"output.file = 'sub/abcd/out.js'",
		"output.assetFileNames = 'assets/sub/[name]_[hash][extname]'",
	].join(" && "),
	() => {
		return outputSnapshotMacro(
			{
				input: "tests/fixtures/src/index1.js",
				plugins: [
					externalAssets("tests/fixtures/assets/*"),
				],
			},
			{
				dir: "sub/abcd",
				assetFileNames: "assets/sub/[name]_[hash][extname]",
			}
		);
	}
);

test(
	[
		"output.dir = 'sub/abcd'",
		"preserveModules = true",
	].join(" && "),
	() => {
		return outputSnapshotMacro(
			{
				input: "tests/fixtures/src/sub/index4.js",
				plugins: [
					externalAssets("tests/fixtures/assets/*"),
				],
			},
			{
				dir: "sub/abcd",
				preserveModules: true,
			}
		);
	}
);

test(
	[
		"multiple inputs with shared chunks",
		"output.chunkFileNames = 'chunks/sub/[name]_[hash].js'",
	].join(" && "),
	() => {
		return outputSnapshotMacro(
			{
				input: ["tests/fixtures/src/index1.js", "tests/fixtures/src/sub/index4.js"],
				plugins: [
					externalAssets("tests/fixtures/assets/*"),
				],
			},
			{
				chunkFileNames: "chunks/sub/[name]_[hash].js",
			}
		);
	}
);

// Will remove empty imports.
test("treeshake.moduleSideEffects = false", () => {
	return outputSnapshotMacro({
		input: "tests/fixtures/src/index1.js",
		plugins: [
			externalAssets("tests/fixtures/assets/*"),
		],
		treeshake: {
			moduleSideEffects: false,
		}
	});
});

test("deduplicate assets with different names", () => {
	return outputSnapshotMacro({
		input: "tests/fixtures/src/index6.js",
		plugins: [
			externalAssets("tests/fixtures/assets/*"),
		],
	});
});
