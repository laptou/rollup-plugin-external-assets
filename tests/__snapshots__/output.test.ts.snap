// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deduplicate assets with different names: index6 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";
import \\"./assets/styles-fc0ceb37.css\\";
import png2 from \\"./assets/image-0fc60877.png\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var main = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

console.log(png2);
main();
"
`;

exports[`multiple inputs with shared chunks && output.chunkFileNames = 'chunks/sub/[name].js': file2 1`] = `
"import \\"../../assets/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

export { sayHi as s };
"
`;

exports[`multiple inputs with shared chunks && output.chunkFileNames = 'chunks/sub/[name].js': index1 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";
import { s as sayHi } from './chunks/sub/file2.js';
import \\"./assets/styles-fc0ceb37.css\\";

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`multiple inputs with shared chunks && output.chunkFileNames = 'chunks/sub/[name].js': index4 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";
import { s as sayHi } from './chunks/sub/file2.js';
import \\"./assets/styles-fc0ceb37.css\\";

var index4 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index4;
"
`;

exports[`output.dir = 'sub/abcd' && output.entryFileNames = 'a/b/[name].js' && output.assetFileNames = 'assets/sub/[name]-[hash][extname]': index1 1`] = `
"import png from \\"../../assets/sub/image-0fc60877.png\\";
import text from \\"../../assets/sub/text-6d7076f2.txt\\";
import \\"../../assets/sub/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`output.dir = 'sub/abcd' && output.entryFileNames = 'a/b/[name].js': index1 1`] = `
"import png from \\"../../assets/image-0fc60877.png\\";
import text from \\"../../assets/text-6d7076f2.txt\\";
import \\"../../assets/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`output.dir = 'sub/abcd' && preserveModules = true: file1 1`] = `
"import text from \\"./assets/text-6d7076f2.txt\\";
export { default } from \\"./assets/text-6d7076f2.txt\\";
"
`;

exports[`output.dir = 'sub/abcd' && preserveModules = true: file2 1`] = `
"import \\"./assets/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

export default sayHi;
"
`;

exports[`output.dir = 'sub/abcd' && preserveModules = true: index4 1`] = `
"import png from \\"../assets/image-0fc60877.png\\";
import text from \\"../assets/text-6d7076f2.txt\\";
import sayHi from '../file2.js';

var index4 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index4;
"
`;

exports[`output.dir = 'sub/abcd': index1 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";
import \\"./assets/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`output.file = 'sub/abcd/out.js' && output.assetFileNames = 'assets/sub/[name]-[hash][extname]': index1 1`] = `
"import png from \\"./assets/sub/image-0fc60877.png\\";
import text from \\"./assets/sub/text-6d7076f2.txt\\";
import \\"./assets/sub/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`output.format = 'amd': index1 1`] = `
"define([\\"./assets/image-0fc60877.png\\", \\"./assets/text-6d7076f2.txt\\", \\"./assets/styles-fc0ceb37.css\\"], function (png, text, styles_css) { 'use strict';

	function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

	var png__default = /*#__PURE__*/_interopDefaultLegacy(png);
	var text__default = /*#__PURE__*/_interopDefaultLegacy(text);

	// This import should not be removed unless specified by the user or some other plugin.

	var sayHi = () => {
		console.log(\\"hello\\");
	};

	var index1 = () => {
		console.log(png__default['default']);
		console.log(text__default['default']);
		sayHi();
	};

	return index1;

});
"
`;

exports[`output.format = 'cjs': index1 1`] = `
"'use strict';

var png = require(\\"./assets/image-0fc60877.png\\");
var text = require(\\"./assets/text-6d7076f2.txt\\");
require(\\"./assets/styles-fc0ceb37.css\\");

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var png__default = /*#__PURE__*/_interopDefaultLegacy(png);
var text__default = /*#__PURE__*/_interopDefaultLegacy(text);

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png__default['default']);
	console.log(text__default['default']);
	sayHi();
};

module.exports = index1;
"
`;

exports[`output.format = 'es': index1 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";
import \\"./assets/styles-fc0ceb37.css\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;

exports[`output.format = 'system': index1 1`] = `
"System.register([\\"./assets/image-0fc60877.png\\", \\"./assets/text-6d7076f2.txt\\", \\"./assets/styles-fc0ceb37.css\\"], function (exports) {
	'use strict';
	var png, text;
	return {
		setters: [function (module) {
			png = module.default;
		}, function (module) {
			text = module.default;
		}, function () {}],
		execute: function () {

			// This import should not be removed unless specified by the user or some other plugin.

			var sayHi = () => {
				console.log(\\"hello\\");
			};

			var index1 = exports('default', () => {
				console.log(png);
				console.log(text);
				sayHi();
			});

		}
	};
});
"
`;

exports[`output.format = 'umd': index1 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require(\\"./assets/image-0fc60877.png\\"), require(\\"./assets/text-6d7076f2.txt\\"), require(\\"./assets/styles-fc0ceb37.css\\")) :
	typeof define === 'function' && define.amd ? define([\\"./assets/image-0fc60877.png\\", \\"./assets/text-6d7076f2.txt\\", \\"./assets/styles-fc0ceb37.css\\"], factory) :
	(global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.test = factory(global.png, global.text));
}(this, (function (png, text) { 'use strict';

	function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

	var png__default = /*#__PURE__*/_interopDefaultLegacy(png);
	var text__default = /*#__PURE__*/_interopDefaultLegacy(text);

	// This import should not be removed unless specified by the user or some other plugin.

	var sayHi = () => {
		console.log(\\"hello\\");
	};

	var index1 = () => {
		console.log(png__default['default']);
		console.log(text__default['default']);
		sayHi();
	};

	return index1;

})));
"
`;

exports[`treeshake.moduleSideEffects = false: index1 1`] = `
"import png from \\"./assets/image-0fc60877.png\\";
import text from \\"./assets/text-6d7076f2.txt\\";

// This import should not be removed unless specified by the user or some other plugin.

var sayHi = () => {
	console.log(\\"hello\\");
};

var index1 = () => {
	console.log(png);
	console.log(text);
	sayHi();
};

export default index1;
"
`;
